### PR TITLE
Block launch if Java version is incompatible

### LINKED ISSUE #54: Check for supported Java version upon launch
Upon launching of an instance:

- Check the Java version being used.
- If the Minecraft version supports it, continue.
- Otherwise:
  * Open a dialog saying that the Java version is not supported.
  * The user may optionally continue on for debugging or experimentation purposes.

MultiMC's support channel is probably 80% people using the wrong Java version. This would fix that.

### LINKED ISSUE #56: newer minecraft versions default to java 17 or show warning.
### Role

playing 1.17 and onward

### Suggestion

newer minecraft versions default to java 17 reducing launch issues

### Benefit

currently you have to change your java version manually to launch 1.17. automatically selecting the latest java will help to remedy this.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

Currently PolyMC users have to manually change java version for any new version of minecraft this is complicated especially for new users. Having the java version automatically switch to java 16 and warning the user if they dont have it installed will reduce confusion. Hopfully if #7 is completed there will be a way to install java 17 automatically after the warning or already have it installed.

### PR DESCRIPTION
**Pull request by Scrumplex**
Closes #54
Closes #56
Requires https://github.com/PolyMC/meta/pull/7

Keep track of compatible Java versions from meta. 
Launch-step `VerifyJavaInstall` will check if current instance's Java version is compatible.
Also add override option both globally and per-instance in-case the user doesn't care about the requirement.

Screenshot

Per-Instance override

Global override

### CODE REVIEW

**Comment by DioEgizio** [comment]
looks like a cool idea but a popup would be better for people who don't check logs

**Comment by DioEgizio** [comment]
also this closes #54

**Comment by Scrumplex** [comment]
> looks like a cool idea but a popup would be better for people who don't check logs

You are kinda forced to check the log anyway. The window will appear if launching fails.

**Comment by DioEgizio** [comment]
yeah but It's easier for some people to have a popup

File: launcher/minecraft/launch/VerifyJavaInstall.cpp
Comment (Scrumplex) [inline launcher/minecraft/launch/VerifyJavaInstall.cpp]: Not happy with this. Maybe we can display it differently?

Meta can suggest multiple versions. For example 1.17-pre1 will recommend `16` and `17`

**Comment by timoreo22** [comment]
Additonnal idea : being able to add an "old" java and "new" java that would be used for corresponding versions automatically instead of having to use per-instance settings

File: launcher/minecraft/launch/VerifyJavaInstall.cpp
Comment (RuinTD) [reply launcher/minecraft/launch/VerifyJavaInstall.cpp]: > Meta can suggest multiple versions. For example 1.17-pre1 will recommend `16` and `17`

I personally wouldn't have it suggest Java 16, since it is discontinued and no longer supported. Security and all that.

**Comment by RuinTD** [comment]
> Additonnal idea : being able to add an "old" java and "new" java that would be used for corresponding versions automatically instead of having to use per-instance settings

#284
Out of scope for this PR

**Review by d-513 — APPROVED** [review]

**Comment by Scrumplex** [comment]
Please ping me before merging this, so I can update meta

File: launcher/minecraft/launch/VerifyJavaInstall.cpp
Comment (Scrumplex) [reply launcher/minecraft/launch/VerifyJavaInstall.cpp]: It's suggesting Java 16, because that's what Mojang is suggesting. In any case this is handled by meta, so if we want to force newer Javas in the future we can totally do that.

**Comment by ZekeZDev** [comment]
> looks like a cool idea but a popup would be better for people who don't check logs

I agree and perhaps once they click ok on the popup it sends them to the java settings.

**Review by txtsd — APPROVED** [review]

**Comment by DioEgizio** [comment]
> > looks like a cool idea but a popup would be better for people who don't check logs
> 
> I agree and perhaps once they click ok on the popup it sends them to the java settings.

yeah many people just refuse to see logs, a popup would be better

**Comment by Scrumplex** [comment]
I just implemented a message box, but I don't really feel comfortable with it, as it doesn't belong into the launchtask at all. If there is any other issue with the instance, it will be reported to log as well.

**Comment by Jan200101** [comment]
Is outright blocking really the right way?

Any experienced power user could end up building their own Java with some custom version number that is interpreted as being incompatible and thus refuse to launch.
How about throwing a warning at the end of a crashed launch or prompting the user with a notice and allowing them to continue?

**Comment by Scrumplex** [comment]
It blocks the launch by default. But you can just skip the verification by turning on a setting, either per-instance or globally (see screenshots in PR description)

Edit: option is called "Skip Java compatibility checks"

**Review by ZekeZDev — APPROVED** [review]

**Comment by Scrumplex** [comment]
Going to merge this once I updated meta

**Comment by jobukkit** [comment]
It thinks Java 11 is incompatible with Minecraft 1.15.2 which is not the case.

**Pull request merged**
